### ISSUE TITLE
Prompt doesn't work properly with non ascii symbols

### ISSUE THREAD

**Issue by kindlycat** [awesomeWM/awesome#974]
For utf-8 characters needs to press "backspace" more than once. 
Navigation by "left" and "right" keys doesn't showing current position, and after navigation "backspace" and "delete" keys doesn't delete anything.

**Comment by psychon** [comment]
Can we make awesome depend on Lua 5.3's UTF support without anyone killing me? Until then, I guess we will have to add more work-arounds like 1cab627d5d48d17f466ce7c58a8932b6b48ca994 to get backspace to work.

For fixing the navigation, I guess we need Pango's help:
https://developer.gnome.org/pango/stable/pango-Layout-Objects.html#pango-layout-get-cursor-pos
https://developer.gnome.org/pango/stable/pango-Layout-Objects.html#pango-layout-move-cursor-visually
(Perhaps Pango can also be used for fixing backspace?)

**Comment by Elv13** [comment]
Dropping 5.1 and luajit support is a big "no!" from me. 5.1 is still by far the most used version of Lua and this is probably never going to change until the Lua comity release a new Lua specification that brings in real value for its core userbase.

As for this issue, it can be fixed using:
https://developer.gnome.org/glib/stable/glib-String-Utility-Functions.html
https://developer.gnome.org/glib/stable/glib-Unicode-Manipulation.html

I didn't check how LGI friendly those functions/macros are. Last time I tried using this API, I got a SIGSEGV, but that was long time ago.

**Comment by SammysHP** [comment]
See also #1420, #1596 and #1744 which are duplicates of this issue.

**Comment by necauqua** [comment]
I've just had same issue while toying with Rust.
Didn't want any complications, so here is what i came up with
``` rust
while bytes.pop().unwrap_or(0) >> 6 == 2 {}
```
I did that because i've noticed that partial unicode bytes start with `10` on the [UTF-8 wiki page](https://en.wikipedia.org/wiki/UTF-8).
Neat little trick, works flawlessly for me.

And also i overwrote `gears.debug.print_error` to make error notifications and it spams those into my face when i'm accidentally on wrong keyboard layout)

**Comment by psychon** [comment]
Note that just removing unicode code points (= ensure the result is valid utf8) is not enough thanks to combining characters.

**Comment by necauqua** [comment]
Well, if you have something like у̃, then you press backspace and it converts to plain y, and then one more to remove the letter - not as bad as pressing backspace 2 times for each of those actions, also spamming warnings. 
Also the 'UGLY HACK' doesn't even work for me btw

**Comment by psychon** [comment]
Yup, that ugly hack seems to have been broken at some point. Back then, widgets were still implemented in C and the C code seems to have thrown a Lua error when `pango_parse_markup` failed. That was broken (by me) in 178204449c477ba437946b94ec29ae019d54caad two and a half years ago. Whoops.

**Comment by necauqua** [comment]
Oh, you were right about surrogate characters, while deleting works fine, idk what to do with selection/movement
But still, operating on code points instead of bytes is much better

**Comment by psychon** [comment]
Newest duplicate (#3308) contains an untested patch that might or might not fix this. The patch needs someone to test it and turn it into a PR.

**Comment by psychon** [comment]
This might be fixed by #3362
